### PR TITLE
micro-ROS handling memory tutorial

### PR DESCRIPTION
**Pull request by pablogs9**
Signed-off-by: Pablo Garrido <pablogs9@gmail.com>

### CODE REVIEW

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+Memory management
+=================
+
+.. contents::
+    :depth: 2
+    :local:
+    :backlinks: none
+
+micro-ROS Client provides a full control over the memory usage during build, configuration and runtime.
+This is one the most important requisites in order to fit in low resources system and also in order to guarantee a hard real-time operation.
```
Comment (dcoello-eProsima) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:12]:
```suggestion
This is one of the most important requisites in order to fit in low resources system and to guarantee a hard real-time behaviour.
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+
+The memory management in micro-ROS Client can be configured at multiple levels so the user can use different mechanisms for fitting its requirements within the micro-ROS environment.
```
Comment (dcoello-eProsima) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:14]:
```suggestion
Memory management in micro-ROS Client can be configured at multiple levels so the user can use different mechanisms for fitting its requirements within the micro-ROS environment.
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+Message memory
+--------------
+
+Message memory handling is an important part of the micro-ROS Client memory handling due to the fact that **micro-ROS Client does not initialized by default the type memory**.
+This means that the user must initialize the type memory before using it.
+This consideration needs to be taken into account both for outgoing and incoming messages.
+
+.. warning::
+
+    micro-ROS provides an API for initializing the type memory that can be found in :ref:`micro-ROS Types Utilities<tutorials_micro_utilities_type_utilities>`.
```
Comment (dcoello-eProsima) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:233]: I think `Note` suits better here than `warning`

Comment (pablogs9) [reply docs/rst/tutorials/micro/memory_management/memory_management.rst:233]: Agree... but I want users to understand that this section is for understanding purposes and that they have the API. I want more than a note but less than a warning

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+Memory management
+=================
+
+.. contents::
+    :depth: 2
+    :local:
+    :backlinks: none
+
+micro-ROS Client provides a full control over the memory usage during build, configuration and runtime.
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:11]:
```suggestion
micro-ROS Client provides full control over the memory usage during build, configuration and runtime.
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+* Static allocated memory in **build time**
+* A bounded stack memory consumption
+* A bounded amount of **dynamic memory** during entity creation and destruction (configuration time)
+* Zero **dynamic memory** during runtime
+
+
+.. TODO(pgarrido): Link to profiling article
+
+
+During a micro-ROS Client application development the user is able configure the memory at multiple level.
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:27]:
```suggestion
During a micro-ROS Client application development the user is able to configure the memory at multiple levels.
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+Along the following sections, those levels are analyzed in detail.
+
+.. _tutorials_micro_memory_management_allocators:
+
+Allocators
+----------
+
+As in the ROS 2 stack, in the micro-ROS stack the dynamic memory allocators can be customized at runtime.
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:35]:
```suggestion
As in the ROS 2 stack, micro-ROS dynamic memory allocators can be customized at runtime.
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+  {
+    // Allocate and return a memory chunk of `number_of_elements * size_of_element` bytes, filled with zeros.
+  }
+
+One example implementation of the most basic allocator that targets platforms where no ``libc`` allocators are available is:
+
+.. note::
+
+  This is a naive implementation of an allocator that does not allows memory deallocation.
+  User shall be aware of using `free_all_heap()` when the micro-ROS entities are no longer required.
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:81]: Maybe place this note below the code?

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+        }
+    }
+
+    #define SYSTEM_ALIGNMENT 4
+
+    size_t align_size(size_t size)
+    {
+        if (size % SYSTEM_ALIGNMENT != 0) {
+            size += SYSTEM_ALIGNMENT - (size % SYSTEM_ALIGNMENT);
+        }
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:107]:
```suggestion
        }
        return size;
```
I guess?

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+
+    // Assigning static memory to the sequence
+    static int32_t memory[100];
+    mymsg.values.capacity = 100;
+    mymsg.values.data = memory;
+    mymsg.values.size = 0;
+
+    // Filling some data
+    for(int32_t i = 0; i < 3; i++){
+        mymsg.values.data = i;
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:322]:
```suggestion
        mymsg.values.data[i] = i;
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+    mypackage__msg__MyComplexType mymsg;
+
+    // Init the multiheaders sequence
+    mymsg.multiheaders.capacity = 10;
+    mymsg.multiheaders.data = (std_msgs__msg__Header*) malloc(mymsg.values.capacity * sizeof(std_msgs__msg__Header));
+    mymsg.multiheaders.size = 0;
+
+    // Filling some data
+    for(int32_t i = 0; i < 3; i++){
+        mymsg.values.data = i;
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:414]:
```suggestion
        mymsg.values.data[i] = i;
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+        // Add memory to this sequence element frame_id
+        mymsg.multiheaders.data[i].frame_id.capacity = 100;
+        mymsg.multiheaders.data[i].frame_id.data = (char*) malloc(mymsg.multiheaders.data[i].frame_id.capacity * sizeof(char));
+        mymsg.multiheaders.data[i].frame_id.size = 0;
+
+        // Assigning value to the frame_id char sequence
+        strcpy(mymsg.multiheaders.data[i].frame_id.data, "Hello World");
+        mymsg.multiheaders.data[i].frame_id.size = strlen(mymsg.multiheaders.data[i].frame_id.data);
```
Comment (Acuadros95) [inline docs/rst/tutorials/micro/memory_management/memory_management.rst:423]: Should this include the null terminator for size?
```suggestion
        mymsg.multiheaders.data[i].frame_id.size = strlen(mymsg.multiheaders.data[i].frame_id.data) + 1;
```

File: docs/rst/tutorials/micro/memory_management/memory_management.rst (hunk 7 above)
Comment (pablogs9) [reply docs/rst/tutorials/micro/memory_management/memory_management.rst]: Fixed here: 9463bbfeeccd4acfd5fba0677ba74ec79a2782d3

File: docs/rst/tutorials/micro/memory_management/memory_management.rst (hunk 3 above)
Comment (pablogs9) [reply docs/rst/tutorials/micro/memory_management/memory_management.rst:234]: Fixed here: 9463bbfeeccd4acfd5fba0677ba74ec79a2782d3

**Review by dcoello-eProsima — APPROVED** [review]

**Review by Acuadros95 — APPROVED** [review]

**Pull request merged**